doc/devs.rst

```diff
@@ -0,0 +1,656 @@
+- ``cannot_remove_peer``
+- ``peer_unavailable``
+
+Please refer to the Doxygen documentation for a detailed explanation as well as
+parameters. The important thing to note is that the peer allows extending its
+basic functionality by extending the callbacks. Also note that we use static
+polymorphism. Any subtype that wishes to extend functionality of the peer
+*hides* the function of its base type and calls the implementation of its base
+type in the function body.
+
+For example, the following code shows how the ``notifier`` extends the
+``peer_connected`` callback:
+
+.. code-block:: C++
+
+  void peer_connected(const peer_id_type& peer_id,
+                      const communication_handle_type& hdl) {
+    BROKER_TRACE(BROKER_ARG(peer_id) << BROKER_ARG(hdl));
+    emit(peer_id, sc::peer_added, "handshake successful");
+    super::peer_connected(peer_id, hdl);
+  }
+
+The ``peer`` is implemented at as template not only because of CRTP, but also to
+allow Broker to configure the types used for the global ID (``PeerId``) and for
+communication handles to other peers (``CommunicationHandle``). The core actor
+sets ``PeerId = caf::node_id`` and ``CommunicationHandle = caf::actor``.
+However, some unit tests use different template parameters.
+
+The member function ``ship`` implements `publishing data`_, but the class
+``peer`` has no code for actually sending messages. The peer leaves this to its
+derived types and requires that ``dref().send(...)`` is well-formed. The core
+actor uses a |alm::stream_transport|_ as communication backend for the
+peer.
+
+Most functions in the ``peer`` are straightforward, but one member function in
+particular is worth discussing:
+
+.. code-block:: C++
+
+  template <class... Fs>
+  caf::behavior make_behavior(Fs... fs)
+
+This member function returns the behavior for initializing the actor that
+implements the peer, i.e., the core actor (leaving unit tests aside). A behavior
+is a set of message handlers (callbacks), usually lambda expressions or `lifted
+member functions <Lifting_>`_. Each class or `mixin <Mixins_>`_ in the
+inheritance graph can add additional message handlers to the actor. Each mixin
+in the "chain" that registers additional message handlers implements
+``make_behavior`` with this exact signature. The template parameter pack
+``fs...`` are the message handler registered by a subtype. Each mixin forwards
+this pack to its base type along with its own handlers.
+
+Each class should document the message handlers it adds to the actor. The sum of
+all message handlers defines the messaging interface of the core actor.
+
+``alm::stream_transport``
+*************************
+
+This class implements a communication backend for |alm::peer|_ that connects two
+actors by using two CAF stream paths (one for each direction, because paths are
+unidirectional).
+
+The stream transport is a CAF `stream manager`_, i.e., it inherits from
+``caf::stream_manager``.
+
+.. _devs.channels:
+
+Channels
```

Comment:
The structure in this section isn't quite clear I think. It's called "Channels" but then both 6.4.1 and 6.4.2 are more specifically talking about how stores use changes (but only 6.4.1 actually says so in its header). How about doing two different subsections instead: 6.4.1 on channels in general, and 6.4.2 on  channels in stores?